### ISSUE TITLE
Enhancement: Add Blazor IBrowserFile.RelativePath to interface

### ISSUE THREAD

**Issue by njannink** [dotnet/aspnetcore#31638]
I would like to be able to access the file(s) that have been selected in the `InputFile` component from javascript to be able to do something like this:

`URL.createobjecturl(file)`

There is a lookup table in the underlying html/javascript element, but the Id of the file is internal. Is it possible to add the Id property to the `IBrowserFile` interface?

Or is there some other way that I can create an extension method for `IBrowserFile` so I can get to the `URL.createobjecturl(file)` underlying javascript call?

**Comment by TanayParikh** [comment]
Hi @njannink, thanks for contacting us. I believe this is a duplicate of https://github.com/dotnet/aspnetcore/issues/27983. I'm closing this issue out in favor of https://github.com/dotnet/aspnetcore/issues/27983 as that already seems to have some community engagements.

**Comment by TanayParikh** [comment]
> I believe this is a duplicate of https://github.com/dotnet/aspnetcore/issues/27983.

Note quite @TanayParikh 😄

**Comment by TanayParikh** [comment]
Resources:
- https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web/src/Forms/InputFile/IBrowserFile.cs
- https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web.JS/src/InputFile.ts#L10-L17

Using this; image previews should be made much simpler:
- https://docs.microsoft.com/en-us/aspnet/core/blazor/images?view=aspnetcore-6.0#preview-an-image-provided-by-the-inputfile-component

**Comment by SteveSandersonMS** [comment]
> There is a lookup table in the underlying html/javascript element

Are you referring to `_blazorFilesById`? If so, that's not public API, so if we were going to have any feature related to this we'd need to make some public API for it.

I'm not sure you can do anything with just the ID information on its own today (without using unsupported private API).

Or am I missing something?

**Comment by TanayParikh** [comment]
> Before this is merged, could you give an example of what code a developer would write to make use of this? I had thought the ID was an internal implementation detail and am not sure what people will be able to do with it.

You're absolutely right that this is an internal implementation detail and it requires access to the internal `_blazorFilesById` map. 

> so if we were going to have any feature related to this we'd need to make some public API for it.

This PR in isolation doesn't immediately facilitate any different flows. Rather, we'll be using this PR / `IBrowserFile.Id` as a building block to achieve https://github.com/dotnet/aspnetcore/issues/27983.

Currently to test it out, we can do something like below:

```csharp
private async Task ResizeAndDisplayImageUsingStreaming(InputFileChangeEventArgs e)
{
    var imageFile = e.File;
    await JS.InvokeVoidAsync("setImage", "showImageHere", imageFile);
}
```

```js
// DO NOT USE THIS AS IT CURRENTLY DEPENDS ON INTERNAL APIs.
​window.setImage = async (imageElementId, imageFile) => {

    const blob = imageElement._blazorFilesById[imageFile.Id].blob;
    const url = URL.createObjectURL(blob);    
    var imageDOMElement = document.getElementById(imageElementDOMId);
    imageDOMElement.src = url;
    URL.revokeObjectURL(url);
  }
```

with https://github.com/dotnet/aspnetcore/issues/27983, we'd ideally be able to do something like:


```js
// FOR ILLUSTRATIVE PURPOSES ONLY, WILL NOT CURRENTLY WORK
​window.setImage = async (imageElementDOMId, imageFile) => {

    const url = imageFile.createObjectURL();    
    var imageDOMElement = document.getElementById(imageElementDOMId);
    imageDOMElement.src = url;
    URL.revokeObjectURL(url);
  }
```

Note; https://github.com/dotnet/aspnetcore/issues/27983 proposes adding the `createObjectURL` to the `IBrowserFile.cs`. Given `createObjectURL` is really a JS / client side construct, I believe it'd be a better fit in `InputFile.ts`, so it can directly leverage the `_blazorFilesById` map. This will allow for easy creation of object URLs using the data which is already client-side. Currently, previewing data from the `InputFile` component [requires a less than ideal round-trip of data](https://docs.microsoft.com/en-us/aspnet/core/blazor/images?view=aspnetcore-6.0#preview-an-image-provided-by-the-inputfile-component). Also note, with the streaming interop, their original request efficient marshaling of file data is already achieved (using `IBrowserFile.OpenReadStream`) should they need the data in .NET.

**Comment by SteveSandersonMS** [comment]
I wonder if there might be a more general-purpose solution here (which would also avoid exposing this implementation detail). Apologies if this has already been considered.

As it is today, in the vast majority of cases, people can already achieve this goal without needing any new API. That is, assuming they have a reference to the `<input type="file">` element (which your code samples above assume), then they can already find the [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) object by looking at `inputFileElem.files` and matching based on the `name` property. We already give the name property value to the .NET code. And in the case of input file elements that don't have `multiple`, they always knows it's `inputFileElem.files[0]`.

So in most cases, if I'm understanding properly, we don't need to do anything to enable this functionality - it's already there in a pretty natural form.

The one case I'm aware of where that approach doesn't quite work is if the user is uploading multiple files, and more than one has an identical name. As far as I know, the only case where this happens is if you put a `webkitdirectory` attribute on the `<input type="file">` so the user gets to upload a complete directory structure, and hence could have multiple files with the same name. However, in that case we already have a functional problem which is that we don't currently expose the relative paths to the files, so the .NET code can't understand the directory structure.

What we could do, then, is instead of adding a new `Id` property, we could add a new property called `RelativePath` which on supported browsers is populated using `webkitRelativePath`. On non-Chromium browsers that don't support uploading whole directories, we'd populate it just using `name` so you always have a valid value. Maybe in the future other browsers will support uploading whole directories and then we'd populate `RelativePath` using whatever info they give us.

If we did this:

 * No need to supply an ID, since the relative path will be equivalent to having an ID for lookup purposes
 * Bonus feature: whole-directory uploads now do something useful, and the .NET-side can preserve the directory structure if you want

AFAIK the implementation cost would actually be less for us to do this, since we literally just send an extra property to .NET, and we're done. We don't need any other APIs for looking up the `File` instances on the JS side.

What do you think? Am I missing anything major?

**Comment by TanayParikh** [comment]
Hey Steve, thanks for the detailed explanation, much appreciated. I think this approach makes a lot of sense, and I especially like the fact we wouldn't need to expose the internal `Id` property.

**Comment by RyanGaudion** [comment]
@TanayParikh or @SteveSandersonMS any plans for this to be developed? My use-case is exactly the one above where I'm using `webkitdirectory` on a file input to be able to upload a directory. 

Currently using `webkitdirectory` and wanting to maintain the same internal folder structure however this seems to be impossible with Blazor at the moment?

**Comment by SteveSandersonMS** [comment]
@RyanGaudion It's just on the backlog right now, which means it's a potential candidate for inclusion in .NET 9 but there's no guarantee yet and there are many other feature candidates that might end up being prioritized more highly. Feedback from you and others does give us justification to focus more on this one, so thanks for commenting.

Since this is an open-source project we are also very willing to work with you if you're interested in submitting a PR. But if you are interested in doing that, please let us know first, and try to provide a design proposal including its pros/cons/alternatives, to maximize the chance that we'd be able to take it.

**Comment by RyanGaudion** [comment]
For those having this issue, I've managed to get a simple workaround working to grab relative paths using JS Interop in C#. Check out the solution [here](https://stackoverflow.com/a/77573286/9100849) on StackOverflow.

Hopefully a more permanent solution will be added in .Net 9.

**Comment by minimalisticMe** [comment]
@SteveSandersonMS Is there an update on this? Did the feature make it into .NET 9?

**Comment by javiercn** [comment]
We'd need to think about this. We could potentially offer a `CreateObjectUrl` method within `IBrowserFile`

**Comment by lvde0** [comment]
@javiercn That would help a lot. Blazor Server sits in a weird spot where files have to be uploaded/streamed to the server first (as `IBrowserFile`) just to be streamed back into the browser to get a blob URL. This roundtrip is quite unnecessary.
